intake_netflow/v9.py

```diff
@@ -0,0 +1,520 @@
+    if flowset_id == 0:
+        return TemplateFlowSet.decode(source)
+    if flowset_id > 255:
+        return DataFlowSet.decode(source)
+    raise Exception("unknown flowset id '{}'".format(flowset_id))
+
+
+class ExportPacket(object):
+    def __init__(self, flowsets, header=None):
+        self.header = header if header else Header(count=len(flowsets))
+        self.flowsets = flowsets
+
+    def update_cache(self, cache):
+        """Update cache of template records."""
+        for flowset in self.flowsets:
+            if not isinstance(flowset, TemplateFlowSet):
+                continue
+            for id, record in flowset.templates.items():
+                cache[id] = record
+
+    def apply(self, templates):
+        """Deserialize partially-decoded data flowsets.
+
+        Deserialization of a data flowset is a two-step process because we
+        cannot assume the needed template is available when we encounter the
+        data flowset. Thus, we place the deserialization process on hold into
+        a packet is fully read. Then we re-scan the partially-decoded data
+        flowsets and finish deserialization.
+        """
+        for i, flowset in enumerate(self.flowsets):
+            if isinstance(flowset, functools.partial):
+                self.flowsets[i] = flowset(templates)
+
+    @staticmethod
+    def decode(source):
+        header = Header.decode(source)
+        flowsets = [decode_flowset(source) for _ in range(header.count)]
+        return ExportPacket(flowsets, header=header)
+
+    def encode(self):
+        raw = self.header.encode()
+        for flowset in self.flowsets:
+            if isinstance(flowset, functools.partial):
+                continue
+            raw += flowset.encode()
+        return raw
+
+
+class PacketStream(object):
+    def __init__(self, source):
+        """A read-only representation of serialized packets.
+
+        Parameters:
+            source : io.BufferedReader
+                Read-only input for packets.
+        """
+        self._source = source
+        self._cache = {}
+
+    def next(self):
+        if self._source.peek() == b'':
+            raise StopIteration
+
+        # Packet deserialization is a two-step process.
+        #
+        # A data flowset requires a defined template to complete the
+        # deserialization process, Since template and data flowsets can be
+        # out-of-order within a packet, then we must temporarily wait to
+        # finish deserializing a data flowset and the respective records.
+
+        packet = ExportPacket.decode(self._source)
+
+        # Add templates to cache
+        packet.update_cache(self._cache)
+
+        # Finish deserialization
+        packet.apply(self._cache)
+
+        return packet
+
+    def __next__(self):
+        return self.next()
+
+    def __iter__(self):
+        return self
+
+    def close(self):
+        return self._source.close()
+
+
+class RecordStream(PacketStream):
```

Comment:
Why the subclass, is PacketStream used directly? Could this be defined as "a bunch of packets read sequentially from a file-like object"?

Reply:
It's not used directly, but it could plausibly be in the future.

The primary reason for two streams is for future maintenance. The packet stream is responsible for reading a packet, caching any templates, and decoding the flowsets. Then the record stream can pull data records out of the decoded packet and flatten the list into a stream. I found separating the logic made it easier for me to reason about during development.